### PR TITLE
⬆️ Upgrade dependencies for 1.8.6

### LINKED ISSUE #1673: ⬆️ Upgrade dependencies
### Related problem

We haven't done a major dependency upgrade since C-PAC v1.8.0

### Proposed feature

Bring dependecies in C-PAC and its image up to date

### Acceptance criteria

- [ ] dependencies updated
- [ ] regression tested

nice to have:
- [ ] #1848
- [ ] CI handles updates on future releases

### Alternatives

_No response_

### Additional context

We can use the random seed feature to help with assessing changes in regression test

### PR DESCRIPTION
**Pull request by shnizzedy**
## Fixes

Resolves #1673 by @shnizzedy 

## Description

* https://github.com/FCP-INDI/C-PAC/blob/6efb38d39c7a40140928810dfe16268a19134ca5/CHANGELOG.md?plain=1#L19-L71
* Fixes CI bugs and improves CI flow.
* Updates dev versioning to match [PEP440](https://peps.python.org/pep-0440/#local-version-segments)

## Technical details

* Refactors workflows from [`workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) triggers to [`workflow_call`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_call) triggers so event data persists through the run
* Adds ability to rebuild specific stages with `[rebuild $STAGE_NAME]` in commit message (with [@amygutierrez](https://github.com/amygutierrez))
  * Now we can see which existing staging images in each stage are determined to be rebuilt, like ![REBUILD: ["standard"]](https://user-images.githubusercontent.com/5974438/228659417-24f84668-8bd3-4790-8dde-ca8c44dfc7e9.png)
  * 32,290 of the added lines are just adding [a copy of `get-pip.py`](dev/docker_data/get-pip_23.0.1.py) to the repo

## Screenshots

<span id="actions" name="actions"/>

### :octocat: Actions before refactor

Check updated version
↓
Build stages
↓
Build and test C-PAC

### :octocat: Actions partially refactored
Build and test C-PAC

### :octocat: Actions refactored
image

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository. 
- [x] My commit messages follow best practices.
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [x] I updated the changelog.
- [x] I added or updated documentation (https://github.com/FCP-INDI/fcp-indi.github.io/pull/300; https://github.com/FCP-INDI/C-PAC/blob/c7fa153f955afa1defc28d652888746edb9c0dfc/.github/README.md).
- [ ] I tried running the project locally and verified that there are no visible errors.

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by shnizzedy** [comment]
Holding back FreeSurfer since we've been talking about pulling it out of the image soon and it's only one version behind.

**Comment by shnizzedy** [comment]
> Why are some miniconda versions being downgraded (to different versions)?

Ah, this branch branched before they were upgraded and I missed the upgrade when backmerging. Restored those versions in 27b9220

File: .github/Dockerfiles/Ubuntu.bionic-non-free.Dockerfile [hunk 1]
```diff
@@ -135,7 +134,6 @@ RUN conda update conda -y && \
       pandas==1.0.5 \
       scipy==1.6.3 \
       traits==4.6.0 \
-      wxpython \
       pip && \
     pip install \
       torch==1.2.0 torchvision==0.4.0 -f https://download.pytorch.org/whl/torch_stable.html && \
```
Comment (nx10) [inline .github/Dockerfiles/Ubuntu.bionic-non-free.Dockerfile:139]: Should  these be updated too? (Or moved/merged with requirements.txt)

Comment (shnizzedy) [reply .github/Dockerfiles/Ubuntu.bionic-non-free.Dockerfile:139]: - The `wxpython` GUI was removed in 1.8.1 and isn't needed for any builds based on this Dockerfile.
- Before this PR, this Ubuntu image was/is the base image for both the standard and ABCD-HCP variant (version-matched to [ABCD-HCP BIDS fMRI Pipeline](https://github.com/DCAN-Labs/abcd-hcp-pipeline/blob/e480a8f99534f1b05f37bf44c64827384b69b383/Dockerfile)) C-PAC images; after/in this PR, [`Ubuntu.bionic-non-free.Dockerfile`](https://github.com/FCP-INDI/C-PAC/blob/27b9220a41df3ce75479943f9a6c9608200f9a24/.github/Dockerfiles/Ubuntu.bionic-non-free.Dockerfile) is still the base for the ABCD-HCP variant, but [`Ubuntu.Python3.10-bionic-non-free.Dockerfile`](https://github.com/FCP-INDI/C-PAC/blob/27b9220a41df3ce75479943f9a6c9608200f9a24/.github/Dockerfiles/Ubuntu.Python3.10-bionic-non-free.Dockerfile) is now the base for the standard image.


Whenever we revisit the variant images, though, we should figure out how to better handle those dependencies that differ between the different variant images.

**Comment by nx10** [comment]
For reference singularity fails on CircleCI with

```bash
SINGULARITYENV_COVERAGE_FILE=.coverage.singularity${VARIANT} singularity exec -B $PWD:/code cpac-singularity-image${VARIANT}.sif /code/dev/circleci_data/test_in_image.sh

FATAL:   container creation failed: mount /proc/self/fd/3->/usr/local/var/singularity/mnt/session/rootfs error: while mounting image /proc/self/fd/3: failed to find loop device: could not attach image file to loop device: no loop devices available

Exited with code exit status 255
```

I tried different versions of singularity but can not get this to go away.
The images are confirmed working locally/manually so this might be a CircleCI issue.

**Comment by amygutierrez** [comment]
When using this image to run C-PAC, I get the following error: 
```
FileNotFoundError: File /usr/share/fsl/6.0/data/atlases/HarvardOxford/HarvardOxford-lateral-ventricles-thr25-2mm.nii.gz does not exist!
```
Looks like that file doesn't exist in `fsl 6.0` 
```
Singularity> pwd
/usr/share/fsl/6.0/data/atlases/HarvardOxford
Singularity> ls
HarvardOxford-cort-maxprob-thr0-1mm.nii.gz   HarvardOxford-cort-maxprob-thr50-1mm.nii.gz  HarvardOxford-sub-maxprob-thr0-1mm.nii.gz   HarvardOxford-sub-maxprob-thr50-1mm.nii.gz  labels
HarvardOxford-cort-maxprob-thr0-2mm.nii.gz   HarvardOxford-cort-maxprob-thr50-2mm.nii.gz  HarvardOxford-sub-maxprob-thr0-2mm.nii.gz   HarvardOxford-sub-maxprob-thr50-2mm.nii.gz  relabel
HarvardOxford-cort-maxprob-thr25-1mm.nii.gz  HarvardOxford-cort-prob-1mm.nii.gz           HarvardOxford-sub-maxprob-thr25-1mm.nii.gz  HarvardOxford-sub-prob-1mm.nii.gz
HarvardOxford-cort-maxprob-thr25-2mm.nii.gz  HarvardOxford-cort-prob-2mm.nii.gz           HarvardOxford-sub-maxprob-thr25-2mm.nii.gz  HarvardOxford-sub-prob-2mm.nii.gz
```

**Comment by shnizzedy** [comment]
> For reference singularity fails on CircleCI with
> 
> ```shell
> SINGULARITYENV_COVERAGE_FILE=.coverage.singularity${VARIANT} singularity exec -B $PWD:/code cpac-singularity-image${VARIANT}.sif /code/dev/circleci_data/test_in_image.sh
> 
> FATAL:   container creation failed: mount /proc/self/fd/3->/usr/local/var/singularity/mnt/session/rootfs error: while mounting image /proc/self/fd/3: failed to find loop device: could not attach image file to loop device: no loop devices available
> 
> Exited with code exit status 255
> ```
> 
> I tried different versions of singularity but can not get this to go away. The images are confirmed working locally/manually so this might be a CircleCI issue.

[Disabled the Singularity tests](https://github.com/FCP-INDI/C-PAC/pull/1916/commits/c7ce58ac8490f0ca1e25bb8961d2be6184070f5d) and made [a new issue](https://github.com/FCP-INDI/C-PAC/issues/1954)

---

> When using this image to run C-PAC, I get the following error:
> 
> ```
> FileNotFoundError: File /usr/share/fsl/6.0/data/atlases/HarvardOxford/HarvardOxford-lateral-ventricles-thr25-2mm.nii.gz does not exist!
> ```
> 
> Looks like that file doesn't exist in `fsl 6.0`
> 
> ```
> Singularity> pwd
> /usr/share/fsl/6.0/data/atlases/HarvardOxford
> Singularity> ls
> HarvardOxford-cort-maxprob-thr0-1mm.nii.gz   HarvardOxford-cort-maxprob-thr50-1mm.nii.gz  HarvardOxford-sub-maxprob-thr0-1mm.nii.gz   HarvardOxford-sub-maxprob-thr50-1mm.nii.gz  labels
> HarvardOxford-cort-maxprob-thr0-2mm.nii.gz   HarvardOxford-cort-maxprob-thr50-2mm.nii.gz  HarvardOxford-sub-maxprob-thr0-2mm.nii.gz   HarvardOxford-sub-maxprob-thr50-2mm.nii.gz  relabel
> HarvardOxford-cort-maxprob-thr25-1mm.nii.gz  HarvardOxford-cort-prob-1mm.nii.gz           HarvardOxford-sub-maxprob-thr25-1mm.nii.gz  HarvardOxford-sub-prob-1mm.nii.gz
> HarvardOxford-cort-maxprob-thr25-2mm.nii.gz  HarvardOxford-cort-prob-2mm.nii.gz           HarvardOxford-sub-maxprob-thr25-2mm.nii.gz  HarvardOxford-sub-prob-2mm.nii.gz
> ```

Rebuilt the staging images, [added a test](https://github.com/FCP-INDI/C-PAC/pull/1916/commits/ade61a39c1e8ed7e0a9c9133fa2fb666dc8e6172) to make sure that file's in each final image at `$FSLDIR/data/atlases/HarvardOxford/HarvardOxford-lateral-ventricles-thr25-2mm.nii.gz` ([it passes](https://output.circle-artifacts.com/output/job/8c15b3fe-2155-4d26-bbdd-874620b8fa71/artifacts/0/htmlcov/d_ea67d2129be80b4a_test_fsl_py.html#t29))

**Pull request merged**